src/compiler/crystal/codegen/primitives.cr

```diff
@@ -173,6 +174,90 @@ class Crystal::CodeGenVisitor
     end
   end
 
+  def codegen_binary_op_add(t : IntegerType, t1, t2, p1, p2)
+    # TODO remove on 0.28
+    return codegen_trunc_binary_op_result(t1, t2, builder.add(p1, p2)) unless @program.has_flag?("preview_overflow")
+
+    llvm_fun = case t.kind
+               when :i8
+                 binary_overflow_fun "llvm.sadd.with.overflow.i8", llvm_context.int8
+               when :i16
+                 binary_overflow_fun "llvm.sadd.with.overflow.i16", llvm_context.int16
+               when :i32
+                 binary_overflow_fun "llvm.sadd.with.overflow.i32", llvm_context.int32
+               when :i64
+                 binary_overflow_fun "llvm.sadd.with.overflow.i64", llvm_context.int64
+               when :i128
+                 binary_overflow_fun "llvm.sadd.with.overflow.i128", llvm_context.int128
+               when :u8
+                 binary_overflow_fun "llvm.uadd.with.overflow.i8", llvm_context.int8
+               when :u16
+                 binary_overflow_fun "llvm.uadd.with.overflow.i16", llvm_context.int16
+               when :u32
+                 binary_overflow_fun "llvm.uadd.with.overflow.i32", llvm_context.int32
+               when :u64
+                 binary_overflow_fun "llvm.uadd.with.overflow.i64", llvm_context.int64
+               when :u128
+                 binary_overflow_fun "llvm.uadd.with.overflow.i128", llvm_context.int128
+               else
+                 raise "unreachable"
+               end
+
+    codegen_binary_overflow_check(llvm_fun, t, t1, t2, p1, p2)
+  end
+
+  # Generates a call to llvm_fun(p1, p2)
+  #
+  # ```
+  # %res_with_overflow = call {T, i1} <llvm_fun>(T %p1, T %p2)
+  # %res = extractvalue {T, i1} %res, 0
+  # %o_bit = extractvalue {T, i1} %res, 1
+  # br i1 %o_bit, label %overflow, label %normal
+  #
+  # overflow:
+  # ;; codegen: raise OverflowError.new with caller's location
+  #
+  # normal:
+  # ;; if T != T1
+  # ;;   %res' is returned
+  # %res' = trunc T %res to T1
+  # ;; else
+  # ;;   %res is returned
+  # ;; end
+  # ```
+  private def codegen_binary_overflow_check(llvm_fun, t : IntegerType, t1, t2, p1, p2)
+    res_with_overflow = builder.call(llvm_fun, [p1, p2])
+
+    res = extract_value res_with_overflow, 0
+    o_bit = extract_value res_with_overflow, 1
+
+    op_overflow = new_block "overflow"
+    op_normal = new_block "normal"
+
+    # TODO raise overflow if outside of t1 type range interpreted as a t2
+
+    cond o_bit, op_overflow, op_normal
```

Comment:
I think we should use `llvm.expect.i1(o_bit, false)` here.

Reply:
Done at "Emit llvm_expect_i1_fun to favor non overflow results"